pkg/clusteragent/admission/mutate/common.go

```diff
@@ -250,15 +245,18 @@ func isApmInstrumentationEnabled(namespace string) bool {
 		return false
 	}
 
-	// Single Step Instrumentation for the namespace can override cluster-wide configuration
-	for _, ns := range apmEnabledNamespaces {
-		if ns == namespace {
-			return true
-		}
+	// Always disable Single Step Instrumentation in kube-system and Datadog namespaces
+	if (namespace == namespaceWithAlwaysDisabledInjections) || (namespace == apiServerCommon.GetResourcesNamespace()) {
```

Comment:
Hi,
I think that defining a function | filter module could be better for code maintainability.

Reply:
I think using filter won't work well in this case. Each filter is not a simple yes or no, but has three states - yes, no or check the next filter. Making this logic to work with filters is possible, but I don't think it will be clean or easy to understand.

Reply:
I follow the point of @AliDatadog, having a function `func filterNamespace(ns string) bool`, can extract from this function the logic from line 243 and 260.

we can even create a struct to encapsulate the exiting generic container filter: https://github.com/DataDog/datadog-agent/blob/c3a2b5547319212e098e53882d2b0e35e088123a/pkg/util/containers/filter.go#L220C19-L220C29

By doing this, it will allow:
* reuse exiting filtering capability, and align the APM admission controller namespace filtering logic with other products.
* regex can be supported easily in the future

```golang
type namespaceFilter struct {
  filter *containers.Filter
}

func newNamespaceFilter(cfg config.Config) (namespaceFilter, error) {
  var includeList, excludeList []string
  // TODO:
  //   - build the exclude and include lists, from the config
  //   - add the prefix `kube_namespace:` to the namespace name in the lists to be compatible
  //     with the containers.Filter logic

  filter, err := containers.NewFilter(containers.GlobalFilter, includeList, excludeList)
  if err != nil {
    return nil, err
  }
  return namespaceFilter{ filter: filter}, nil
}

func (f* namespaceFilter) filter(namespace string) bool {
  return !f.filter.IsExcluded(nil,"","",namespace)
}
```

So it is maybe not necessary for this PR. but it can be a good refactoring

Reply:
We need to support three checks for the namespace value:
1. `if len(includeList) > 0 && len(excludeList) > 0` - this cannot be checked by filter. From the filter point of view, it's valid to have both includeList and excludeList set at the same time. So we cannot replace this check by the filter.
2.` if (namespace == "kube_system") || (namespace == "datadog")` - we could add kube_system and datadog namespaces to the excludeList; however since includeList takes priority, users can add kube_system to the includeList and do APM Instrumentation in Kubernetes namespace. That goes against expected feature behavior. 
3. 	```
         if len(includeList) > 0 {
		if slices.Contains[[]string, string](includeList, namespace) {
			return true
		}
		return false
	}
        if slices.Contains[[]string, string](excludeList, namespace) {
		return false
	}
```
I think this logic can be _almost_ replaced by **!f.filter.IsExcluded(nil,"","",namespace)**, except for returning false if includeList is non-empty, but doesn't contain current namespace. Maybe I'm missing how it can be done?

I could see the filtering logic getting complicated enough to warrant using filters. But, I cleaned up the logic a bit and now it's super simpler and readable. I think it will be clearer to our future selves especially those without knowledge of the filter API, to use the current solution. Let me know what you think

Reply:
* for point `1.` : this can be validated at the admission controller initialisation once, and not each time the admission controller process a request. like this it is not anymore the responsibility of the filter to check a miss configuration. Also I think we should define a priority between `enabled/disabled` so even with a `bad` configuration the admission controller is working see my other comment: https://github.com/DataDog/datadog-agent/pull/21673#discussion_r1441499559

* for point `2.`: IMO we can protect the customer from bad configuration by excluding `kube-system` by default, but if the customer set it on purpose, it is his responsibility, and we should "accept that". We do the same in the agent with "pause container", we exclude them by default, but if a customer see a good reason to not exclude them, they can configure the agent to do it.

* for point `3.`:  the `filter.IsExcluded` function return alway `false` is the `NamespaceIncludeList` is empty. So I think it work as expected see: https://github.com/DataDog/datadog-agent/blob/1f3aa6a2eaeb865046ee959f0572098f3759783c/pkg/util/containers/filter.go#L348C23-L348C43

Reply:
1. 👍 for moving this check to run once when admission controller starts
2. I do think that instrumenting kube-system could be too dangerous even if users explicitly want it. But I'm OK to do it if it's OK from product POV
3.  If includeList is non-empty, but doesn't contain current namespace, filter.IsExcluded will return false - https://github.com/DataDog/datadog-agent/blob/1f3aa6a2eaeb865046ee959f0572098f3759783c/pkg/util/containers/filter.go#L379. But we want the opposite - if includeList is non-empty and the current namespace is not in this list, the namespace should be excluded, i.e. IsExcluded should return true.